File: testehash.c
/*
 * O testehash é uma ferramenta simples para geração de hash.
 * Desenvolvi para demonstrar a produção de hashs em sistemas e criptomoedas
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <openssl/sha.h> //Apesar de existirem diversas bibliotecas criptográficas, a libssl é uma das mais usadas
                         //a documentação é, infelizmente, uma porcaria

//A função digest recebe uma string, como array de chars e retorna o hexdigest (saida hexadecimal em tamanho duplo)
//como string (ponteiro de char)
char* digest(char mensagem[]){

    char *strout = (char*)malloc(sizeof(char)*SHA256_DIGEST_LENGTH);//A string de saída é alocada com o tamanho padrão da
                                                                    //SHA256

    unsigned char hash[SHA256_DIGEST_LENGTH];//A hash é armazenada em estado bruto num array de chars sem sinal 
    /* O algoritmo para encriptar a mensagem, de forma mais comum é:
     * Criar um "wrapper" contexto. O contexto é um conjunto de dados previamente gerados que será usado pela função para produzir o hash
     * O contexto é inicializado, atualizado com os dados da mensagem e por fim o hash é produzido
     */
    SHA256_CTX contexto;
    SHA256_Init(&contexto);
    SHA256_Update(&contexto, mensagem, strlen(mensagem));
    SHA256_Final(hash, &contexto);

    /* Como o hash é um array de tamanho fixo (SHA256_DIGEST_LENGTH), contendo em seus elementos os trechos do hash, não é possível
     * imprimir diretamente seus dados. Pra isso usamos o seguinte algoritmo:
     * Usamos uma variável temporária (buff) do tipo char, com tamanho "duplo", para receber cada um dos itens. 
     * Como não ha metodo direto para isso, mas há como imprimir com a string de formato 02x (hexadecimal de 2 posições, completados com zeros a esquerda)
     * usamos a função sprintf para imprimir para a variável temporária buff 
     * O conteúdo de buff é concatenado na variável de saída strout, que é retornada
     */

    for(int i = 0; i < SHA256_DIGEST_LENGTH; i++){
        char *buff = (char*)malloc(sizeof(char)*2+1);

        sprintf(buff, "%02x", hash[i]);
        strcat(strout, buff);

    }

    return strout;
}

//A função imprime o hash de um inteiro. O inteiro é primeiro "impresso para uma variável" string, 
//já que os resultados para o tratamento para uma string e para os números são diferentes (a codificação varia)
void _imp_int(int c){
    char *_c;
    _c = (char*)malloc(sizeof(char)*SHA256_DIGEST_LENGTH);

    sprintf(_c, "%d", c);
    printf("| Inteiro: %08d | Hash: %s\n", c, digest(_c));
    free(_c);

}

/* A funçao gera o hahs da string recebida. 
 * O condicional e meramente estetico, permitindo que strings muito extensas sejam impressas numa linha e seu hash em ou-
 * tra. 
 */
void _imp_str(char *c){

    if(strlen(c)<16)
    	printf("| String: %s | Hash: %s\n", c, digest(c));
    else
	printf("| String: %s \n| Hash: %s\n", c, digest(c));
}

int main(int argc, char *argv[]){

    srand(time(NULL)); //Inicializa a semente de números pseudoaleatórios, usando o tempo do sistema
		       //Felizmente este não é um sistema "seguro"

    /*Existem interpretadores de argumentos em C, mas não são simples de usar e criam dependências externas 
     *A abordagem mais simples é só procurar na linha de comando em posições específicas pelos argumentos esperados 
     *O programa, no entanto, perde a flexibilidade
     *No nosso caso contamos a quantidade de argumentos e comparamos a segunda posição pelos argumentos n (para gerar números 
     *aleatórios), s (para encriptar uma string) ou sl (para encadear a sequencia de stringis a seguir e processar como uma string única)
     */
    if(argc==3 && strcmp(argv[1], "-n")==0){
	char *ptr;
	long int v = strtol(argv[2], &ptr, 10); /*A função strtol separa letras e números de uma string, colocando os dados alfanuméricos no ponteiro ptr 
						 *e o número existe é retornado (armazenado na variável v, no nosso caso). Caso não existam números (ou ele
						 *não esteja antes das letras) a variável v tem valor 0 e o programa sugere ao usuário o uso correto. 
						 */
	if(v==0)
            printf("Digite um número inteiro para informar a quantidade de hashs de números inteiros a serem gerados\n");
        else
            for(int i=0; i<atoi(argv[2]); i++)
                _imp_int(rand()%99999999); // Número aleatório de 8 posições 
    }

    //Neste caso tudo que estiver depois do argumento -s é tratado como string individual e seu hash gerado individualmente
    else if(argc>=3 && strcmp(argv[1], "-s")==0){
        for(int i=2; i<argc; i++)
            _imp_str(argv[i]);
    }

    //Neste caso (-sl) as strings apresentadas após o argumento são concatenadas e encriptadas como uma string só
    else if(argc>=3 && strcmp(argv[1], "-sl")==0){

        char buffer[1024];

        for(int i=2; i<argc; i++){
	//O condicional abaixo evita que os espaços entre as strings sejam ignorados ou que um espaço seja incluído no começo da string
	//(o que fatalmente resultaria o hash de saída)
            if(strcmp(buffer, "")==0)
                strcat(buffer, argv[i]);
            else {
                strcat(buffer, " ");
                strcat(buffer, argv[i]);
            }
        }

        _imp_str(buffer);

    }
    else {
        printf("Uso do testehash:\n");
        printf("testehash -n X - O programa gerará X números aleatórios e seus respectivos hashes\n");
        printf("testehash -s <string1> ... <stringX> - O programa apresentará os hashs das X strings apresentadas \n");
        printf("testehash -sl <string> - O programa unirá todas as strings apresentadas e criará um hash único\n");

    }

}

File: testehash.py
#!/usr/local/bin/python
# coding: utf8


import hashlib #Funçoes criptograficas do python
import random #Criaçao de numeros pseudoaleatorios
import sys #Permite usar a interface com o sistema para ler as linhas de comandos

from argparse import ArgumentParser
#Permite usar o processador de argumentos para geraçao da ajuda tratamento automatizado de argumentos da linha de comando

"""
A funçao abaixo cria o objeto criptografico strout, com algoritmo sha256
A mensagem recebia como parametro e convertida em string (mesmo que ja seja)
E retornado o hash tratado como hexadecimais de tamanho duplo (hexdigest) 
"""
def digest(mensagem):
    strout = hashlib.sha256()
    strout.update(str(mensagem))
    return strout.hexdigest()


"""
A funçao gera n numeros aleatorios e seus hashs, usando a funçao digest acima
Os numeros gerados sao inteiros de 24 bits (ate 8 digitos), formatados com zeros 
iniciais (08d)
"""
def _imp_int(n):
    i = 0
    while i < n:
        x = random.getrandbits(24)
        print('| Inteiro: {0:08d} | Hash: {1}'.format(x, digest(x)))
        i += 1

"""
A funçao gera o hahs da string recebida. 
O condicional e meramente estetico, permitindo que strings muito extensas sejam impressas numa linha e seu hash em ou-
tra. 
"""
def _imp_str(s):
    if len(s)<16:
        print('| String: {0} | Hash: {1}'.format(s, digest(s)))
    else:
        print('| String: {0} \n| Hash: {1}'.format(s, digest(s)))


#Main
"""
O processador de argumentos e primeiro criado e as 3 funçoes que criei na versao original em java e implementei na 
versao em C
As funçoes possuem arguumentos compostos (nargs = '+') podem receber varias strings na linha de comando sem 
a necessidade de tratamentos extras no codigo 
"""
parser = ArgumentParser(description="Simula a geração de hashs a partir de inteiros aleatórios ou strings")
parser.add_argument("-n", help="O programa gerará X números aleatórios e seus respectivos hashes", type=int,
                    metavar='X', dest='X')
parser.add_argument("-s", help="O programa apresentará os hashs das N strings apresentada", type=str,
                    metavar='String1, String2, ... StringN', nargs='+')
parser.add_argument("-sl", help="O programa unirá todas as strings apresentadas e criará um hash único", type=str,
                    metavar='S', nargs='+')
parser.parse_args() #processa os argumentos

"""
sys.argv e uma lista, que funciona de forma similar aos arrays argv do java e C, com a facilidade da busca automatica
"""
if ("-n" in sys.argv):
    _imp_int(int(sys.argv[sys.argv.index("-n") + 1]))
if "-s" in sys.argv:
    for i in range(sys.argv.index("-s") + 1, len(sys.argv)):
        _imp_str(sys.argv[i])
""" Nao ha mecanismo similar a stringbuilders em python. Pra contornar essa situaçao, criei uma lista chamada buff
    que anexa todas as strings passadas na linha de comando e que depois e unida (''.join). A string resultante e
    entao encripada usando a funçao _imp_str."""
if "-sl" in sys.argv:
    buff = []
    for i in range(sys.argv.index("-sl") + 1, len(sys.argv)):
        if len(buff) == 0:
            buff.append(sys.argv[i])
        else:
            buff.append(" ")
            buff.append(sys.argv[i])
    _imp_str(''.join(buff))